### PR TITLE
adding {u,d,c,s,b}->4l to JHUGenMELA through Madgraph

### PR DESCRIPTION
**Pull request by MohitS704**
This PR adds 3 new processes to JHUGenMELA utilizing the Madgraph Matrix Elements introduced in #52. 

light-quark (u,d,c,s,b) -> 4l signal, background, and (signal+background+interference). A plot of a Madgraph BSI sample reweighted to other terms using MELA is depicted below:

image

Infrastructure for making the Madgraph matrix elements work for MELA has also been introduced in the form of a map between compute/update functions and the process/production TVars.

Three new processes were added into MELA to encompass the madgraph distinctions, as well as a new production mode called qqH.

### CODE REVIEW

File: MELA/src/MadMela.cc [hunk 1]
```diff
@@ -8,6 +8,8 @@
 #include "TMCFM.hh"
 using MELAStreamHelpers::MELAout;
 using MELAStreamHelpers::MELAerr;
+using namespace madMela;
+std::map<std::pair<TVar::Process, TVar::Production>, MG_process_double>* madMela::updateMap;
```
Comment (MohitS704) [inline MELA/src/MadMela.cc:12]: This map serves as a quick way to swap between compute functions (faster than a series of if statements) for different productions/processes

File: MELA/src/TUtil.cc [hunk 2]
```diff
@@ -1715,6 +1715,8 @@ double TUtil::InterpretScaleScheme(const TVar::Production& production, const TVa
         production == TVar::ttH
         ||
         production == TVar::bbH
+        ||
+        production == TVar::qqH
```
Comment (MohitS704) [inline MELA/src/TUtil.cc:1719]: implement the same scaling as heavy quarks for light quarks (because why not I guess).

File: MELA/src/TUtil.cc [hunk 3]
```diff
@@ -5271,20 +5273,22 @@ double TUtil::MadgraphMatEl(
...
-
   TLorentzVector MomStore[mxpart]; // Mom (in natural units) to compute alphaS
   for (int i = 0; i < mxpart; i++) MomStore[i].SetXYZT(0, 0, 0, 0);
-  // if(verbosity >= TVar::DEBUG) MELAout << "Boost vector of " << boostVec.Px() << " " << boostVec.Py() << " " << boostVec.Pz() << endl;
+
+  // vector<int> pdgs(nPDG);
+  // vector<vector<double>> p(nPDG, vector<double>(4));
+  
+  int pdgs[nPDG];
+  double* p = new double[4*nPDG];
```
Comment (MohitS704) [inline MELA/src/TUtil.cc:5283]: Overhaul the creation of the input 4-vectors to skip the first vector-quantity part to increase speed

File: MELA/src/TUtil.cc [hunk 4]
```diff
@@ -5355,26 +5350,32 @@ double TUtil::MadgraphMatEl(
...
     for(int i = 0; i < nPDG; i++){
       for(int j = 0; j < 4; j++){
-        MELAout << p_for_fortran[i*4+j] << " ";
+        MELAout << p[i*4+j] << " ";
       }
       MELAout << endl;
     }
   }
-  madMela::update_all_coup_();
+  madMela::update_all_coup(process, production);
```
Comment (MohitS704) [inline MELA/src/TUtil.cc:5368]: `madMela::update_all_coup` now requires the process and the production to correctly compute what it needs

File: MELA/interface/MadMela.h [hunk 5]
```diff
@@ -2,53 +2,159 @@
...
+            gc_1312,gc_1313,gc_1314,gc_1315,gc_1316,gc_1317,gc_1318,gc_1355,
+            gc_1356,gc_1360,gc_1364,gc_1365,gc_1366,gc_1367,gc_1374,gc_1375,
+            gc_1382,gc_1383,gc_1387,gc_1391,gc_1392,gc_1393,gc_1394,gc_1401,
+            gc_1402;
         }couplings_; //you shouldn't have to edit these, but they are here for safekeeping!
 
-        void smatrixhel_(int pdgs[], int& procid, const int& npdg, double* p, double& alphas, double& scale2, int& nhel, double& ans);
-        void update_all_coup_();
-        void get_pdg_order_(int pdg[6][3], int allproc[3]);
+        void ggFSIG_smatrixhel_(int pdgs[], int& procid, const int& npdg, double* p, double& alphas, double& scale2, int& nhel, double& ans);
```
Comment (MohitS704) [inline MELA/interface/MadMela.h:126]: all new shiny compute functions!

**Comment by MohitS704** [comment]
The full area (that will be placed in the JHUGen madMELA folder for compilation) exists within this repo:

https://github.com/hexutils/madMELA_storage/tree/main

I will clean it up then move forwards to including that as a JHUGen PR

**Comment by MohitS704** [comment]
These newest commits deal with adding taus into the probabilities for reweighting

**Review by gritsan — APPROVED** [review]

**Review by usarica — CHANGES_REQUESTED** [review]
I do no understand the plot - do you apply any cuts on kinematics? What happens if you do not? Are there any implicit cuts on bkg MEs?

Also, on the choice of the production name, there is an existing production mode for this, ZZQQB. I would suggest to keep that instead of creating a new production mode. The background component should match MCFM with process bkgZZ. Do you have any event-by-event validation of the MEs?

**Comment by MohitS704** [comment]
> I do no understand the plot - do you apply any cuts on kinematics? What happens if you do not? Are there any implicit cuts on bkg MEs?
> 
> Also, on the choice of the production name, there is an existing production mode for this, ZZQQB. I would suggest to keep that instead of creating a new production mode. The background component should match MCFM with process bkgZZ. Do you have any event-by-event validation of the MEs?

The plot is just BSI generated with madgraph then subsequently reweighted with madgraph. I'll attach the run card to the next comment.

I do have event-by-event validation here:
image

For all intents and purposes it is close enough, imo.

Also I changed it so that the TVar is uses is ZZQQB.

**Comment by MohitS704** [comment]
#*********************************************************************
#                       MadGraph5_aMC@NLO                            *
#                                                                    *
#                     run_card.dat MadEvent                          *
#                                                                    *
#  This file is used to set the parameters of the run.               *
#                                                                    *
#  Some notation/conventions:                                        *
#                                                                    *
#   Lines starting with a '# ' are info or comments                  *
#                                                                    *
#   mind the format:   value    = variable     ! comment             *
#                                                                    *
#   To display more options, you can type the command:               *
#      update to_full                                                *
#*********************************************************************
#                                                                    
#*********************************************************************
# Tag name for the run (one word)                                    *
#*********************************************************************
  tag_1     = run_tag ! name of the run 
#*********************************************************************
# Number of events and rnd seed                                      *
# Warning: Do not generate more than 1M events in a single run       *
#*********************************************************************
  100000 = nevents ! Number of unweighted events requested 
  0   = iseed   ! rnd seed (0=assigned automatically=default))
#*********************************************************************
# Collider type and energy                                           *
# lpp: 0=No PDF, 1=proton, -1=antiproton, 2=photon from proton,      *
#                3=photon from electron, 4=photon from muon          *
#*********************************************************************
     1        = lpp1    ! beam 1 type 
     1        = lpp2    ! beam 2 type
     6500.0     = ebeam1  ! beam 1 total energy in GeV
     6500.0     = ebeam2  ! beam 2 total energy in GeV
# To see polarised beam options: type "update beam_pol"

#*********************************************************************
# PDF CHOICE: this automatically fixes also alpha_s and its evol.    *
#*********************************************************************
     nn23lo1    = pdlabel     ! PDF set                                     
     230000    = lhaid     ! if pdlabel=lhapdf, this is the lhapdf number
# To see heavy ion options: type "update ion_pdf"
#*********************************************************************
# Renormalization and factorization scales                           *
#*********************************************************************
 False = fixed_ren_scale  ! if .true. use fixed ren scale
 False        = fixed_fac_scale  ! if .true. use fixed fac scale
 91.188  = scale            ! fixed ren scale
 91.188  = dsqrt_q2fact1    ! fixed fact scale for pdf1
 91.188  = dsqrt_q2fact2    ! fixed fact scale for pdf2
 -1 = dynamical_scale_choice ! Choose one of the preselected dynamical choices
 1.0  = scalefact        ! scale factor for event-by-event scales
#*********************************************************************
# Type and output format
#*********************************************************************
  False     = gridpack  !True = setting up the grid pack
  -1.0 = time_of_flight ! threshold (in mm) below which the invariant livetime is not written (-1 means not written)
  average =  event_norm       ! average/sum. Normalization of the weight in the LHEF
# To see MLM/CKKW  merging options: type "update MLM" or "update CKKW"

#*********************************************************************
#
#*********************************************************************
# Phase-Space Optimization strategy (basic options)
#*********************************************************************
   0  = nhel          ! using helicities importance sampling or not.
                             ! 0: sum over helicity, 1: importance sampling
   1  = sde_strategy  ! default integration strategy (hep-ph/2021.00773)
                             ! 1 is old strategy (using amp square)
			     ! 2 is new strategy (using only the denominator)
# To see advanced option for Phase-Space optimization: type "update psoptim"			     
#*********************************************************************
# Generation bias, check the wiki page below for more information:   *
#  'cp3.irmp.ucl.ac.be/projects/madgraph/wiki/LOEventGenerationBias' *
#*********************************************************************
 None = bias_module  ! Bias type of bias, [None, ptj_bias, -custom_folder-]
 {} = bias_parameters ! Specifies the parameters of the module.
#
#*******************************                                                 
# Parton level cuts definition *
#*******************************                                     
#                                                                    
#
#*********************************************************************
# BW cutoff (M+/-bwcutoff*Gamma) ! Define on/off-shell for "$" and decay  
#*********************************************************************
  15.0  = bwcutoff      ! (M+/-bwcutoff*Gamma)
#*********************************************************************
# Standard Cuts                                                      *
#*********************************************************************
# Minimum and maximum pt's (for max, -1 means no cut)                *
#*********************************************************************
 -1.0  = ptl       ! minimum pt for the charged leptons 
 -1.0  = ptlmax    ! maximum pt for the charged leptons
 {} = pt_min_pdg ! pt cut for other particles (use pdg code). Applied on particle and anti-particle
 {}	= pt_max_pdg ! pt cut for other particles (syntax e.g. {6: 100, 25: 50}) 
#
# For display option for energy cut in the partonic center of mass frame type 'update ecut'
#
#*********************************************************************
# Maximum and minimum absolute rapidity (for max, -1 means no cut)   *
#*********************************************************************
 -1  = etal    ! max rap for the charged leptons 
 0.0  = etalmin ! main rap for the charged leptons
 {} = eta_min_pdg ! rap cut for other particles (use pdg code). Applied on particle and anti-particle
 {} = eta_max_pdg ! rap cut for other particles (syntax e.g. {6: 2.5, 23: 5})
#*********************************************************************
# Minimum and maximum DeltaR distance                                *
#*********************************************************************
 -1.0 = drll    ! min distance between leptons 
 -1.0  = drllmax ! max distance between leptons
#*********************************************************************
# Minimum and maximum invariant mass for pairs                       *
 # WARNING: for four lepton final state mmll cut require to have      *
 #          different lepton masses for each flavor!                  *           
#*********************************************************************
 4.0   = mmll    ! min invariant mass of l+l- (same flavour) lepton pair
 -1.0  = mmllmax ! max invariant mass of l+l- (same flavour) lepton pair
 {} = mxx_min_pdg ! min invariant mass of a pair of particles X/X~ (e.g. {6:250})
 {'default': False} = mxx_only_part_antipart ! if True the invariant mass is applied only 
                       ! to pairs of particle/antiparticle and not to pairs of the same pdg codes.  
 #*********************************************************************
 # Minimum and maximum invariant mass for all letpons                 *
 #*********************************************************************
 0.0   = mmnl    ! min invariant mass for all letpons (l+- and vl) 
 -1.0  = mmnlmax ! max invariant mass for all letpons (l+- and vl) 
 #*********************************************************************
 # Minimum and maximum pt for 4-momenta sum of leptons / neutrino     *
 #  for pair of lepton includes only same flavor, opposite charge
 #*********************************************************************
 0.0   = ptllmin  ! Minimum pt for 4-momenta sum of leptons(l and vl)
 -1.0  = ptllmax  ! Maximum pt for 4-momenta sum of leptons(l and vl)
#*********************************************************************
# Inclusive cuts                                                     *
#*********************************************************************
 0.0  = xptl ! minimum pt for at least one charged lepton 
 #*********************************************************************
 # Control the pt's of leptons sorted by pt                           *
 #*********************************************************************
 0.0   = ptl1min ! minimum pt for the leading lepton in pt
 0.0   = ptl2min ! minimum pt for the second lepton in pt
 0.0   = ptl3min ! minimum pt for the third lepton in pt
 0.0   = ptl4min ! minimum pt for the fourth lepton in pt
 -1.0  = ptl1max ! maximum pt for the leading lepton in pt 
 -1.0  = ptl2max ! maximum pt for the second lepton in pt
 -1.0  = ptl3max ! maximum pt for the third lepton in pt
 -1.0  = ptl4max ! maximum pt for the fourth lepton in pt
#*********************************************************************
# maximal pdg code for quark to be considered as a light jet         *
# (otherwise b cuts are applied)                                     *
#*********************************************************************
 5 = maxjetflavor    ! Maximum jet pdg code
#*********************************************************************
#
#*********************************************************************
# Store info for systematics studies                                 *
# WARNING: Do not use for interference type of computation           *
#*********************************************************************
   True  = use_syst      ! Enable systematics studies
#
systematics = systematics_program ! none, systematics [python], SysCalc [depreceted, C++]
['--mur=0.5,1,2', '--muf=0.5,1,2', '--pdf=errorset'] = systematics_arguments ! see: https://cp3.irmp.ucl.ac.be/projects/madgraph/wiki/Systematics#Systematicspythonmodule
# Syscalc is deprecated but to see the associate options type'update syscalc'

**Review by gritsan — APPROVED** [review]

**Comment by MohitS704** [comment]
@usarica I believe I've answered your questions on this

**Review by gritsan — APPROVED** [review]

**Review by jdavis36 — DISMISSED** [review]

**Review by gritsan — APPROVED** [review]

File: MELA/interface/TVar.hh [hunk 6]
```diff
@@ -182,6 +180,10 @@ namespace TVar{
     SelfDefine_spin1,
     SelfDefine_spin2,
 
+    /** Madgraph Signal, Background, BSI **/
+    Madgraph_sig,
+    Madgraph_bkg,
+    Madgraph_bsi,
```
Comment (usarica) [inline MELA/interface/TVar.hh:186]: Why do you need separate processes for MG? Aren't HSMHiggs, bkgZZ, and bkgZZ_SMHiggs enough?

File: MELA/python/mela_binding.cpp [hunk 7]
```diff
@@ -221,13 +221,13 @@ SimpleParticleCollection_t collection_initializer(py::list listOfParticles){
 #define MAKE_COUPLING_ARR_SPIN_ZERO(arrayName, size, arrType)\
         .def(#arrayName, [](py::object &obj){ \
             Mela &D = obj.cast<Mela&>(); \
-            return py::array_t<arrType>(std::vector<int>{nSupportedHiggses, size, 2}, (const arrType*) &D.arrayName, obj); \
+            return py::array_t<arrType>(std::vector<arrType>{nSupportedHiggses, size, 2}, (const arrType*) &D.arrayName, obj); \
```
Comment (usarica) [inline MELA/python/mela_binding.cpp:224]: Would you mind reminding me why arrType is needed?

**Review by usarica — CHANGES_REQUESTED** [review]

File: MELA/interface/TVar.hh (hunk 6 above)
Comment (MohitS704) [reply MELA/interface/TVar.hh:186]: I guess this is true - but there are some stark differences between madgraph and JHUGen/MCFM elements. For instance, everything is based off of the SMEFTsim calculation base parameters (so changing the higgs mass would change the Z or W mass, for instance), which is why I wanted to emphasize that with a different process.

If you think that the different matrix element is enough to ensure the user doesn't do something unwanted then yeah I can change it

Comment (usarica) [reply MELA/interface/TVar.hh:186]: I think a different ME setting should be enough - people should be able to keep track of different MELA behavior from the documentation.

Comment (MohitS704) [reply MELA/interface/TVar.hh]: Done

**Review by gritsan — APPROVED** [review]

**Review by usarica — APPROVED** [review]

**Pull request merged**